pkg/resource/defaultclass.go

```diff
@@ -0,0 +1,133 @@
+
+	// Create list options query that will be used to search
+	// for resource class that is default for claim kind.
+	options := &client.ListOptions{}
+	if err := options.SetLabelSelector(gk + "/default=" + "true"); err != nil {
+		// Panic if unable to set label selector or else panic will occur
+		// when returned reconciler is invoked.
+		panic(err)
+	}
+
+	return &DefaultClassReconciler{
+		client:   m.GetClient(),
+		newClaim: nc,
+		options:  options,
+	}
+}
+
+// Reconcile reconciles a claim to the default class reference for its kind
+func (r *DefaultClassReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	logDefaultClass.V(logging.Debug).Info("Reconciling", "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	claim := r.newClaim()
+	if err := r.client.Get(ctx, req.NamespacedName, claim); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetClaim)
+	}
+
+	// Get resource classes with claim kind as default
+	classes := &corev1alpha1.ResourceClassList{}
+	if err := r.client.List(ctx, r.options, classes); err != nil {
+		// If this is the first time we encounter no defaults we'll be
+		// requeued implicitly due to the status update. If not, we don't
+		// care to requeue because list parameters will not change.
+		claim.SetConditions(corev1alpha1.ReconcileError(errors.New(errFailedList)))
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errUpdateClaimStatus)
+	}
+
+	// Check to see if no defaults defined for claim kind.
+	if len(classes.Items) == 0 {
+		// If this is the first time we encounter no defaults we'll be
+		// requeued implicitly due to the status update. If not, we will requeue
+		// after a time to see if a default class has been created.
+		claim.SetConditions(corev1alpha1.ReconcileError(errors.New(errNoDefaultClass)))
+		return reconcile.Result{RequeueAfter: defaultClassWait}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errUpdateClaimStatus)
+	}
+
+	// Check to see if multiple defaults defined for claim kind.
+	if len(classes.Items) > 1 {
+		// If this is the first time we encounter multiple defaults we'll be
+		// requeued implicitly due to the status update. If not, we will requeue
+		// after a time to see if only one default class exists.
+		claim.SetConditions(corev1alpha1.ReconcileError(errors.New(errMultipleDefaultClasses)))
+		return reconcile.Result{RequeueAfter: defaultClassWait}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errUpdateClaimStatus)
+	}
+
+	// Set class reference on claim to default resource class
+	claim.SetClassReference(meta.ReferenceTo(classes.Items[0].GetObjectMeta(), corev1alpha1.ResourceClassGroupVersionKind))
```

Comment:
I don't think you need the `GetObjectMeta` call here - I believe `classes.Items[0]` (which I believe is a `*corev1alpha1.ResourceClass`) satisfies the `metav1.Object` interface.